### ISSUE TITLE
[dev,train-2013.03.28] expired certificates are not being invalidated.

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#3153]
STR
1. Start with a brand new profile.
2. Go to dev.123done.org, and use the dialog to signin with an @yahoo.com email address.
3. After the dialog dismisses, do nothing for >5 minutes.
4. Then logout of dev.123done.org and try to signin again with that @yahoo.com address.

Expected: Signed in.
Actual: alert is shown with 'failed to verify assertion: assertion has expired'.

Basically from looking at localStorage, the certificate for that email is not invalidated when it has expired. And then that expired certificate is used for the assertion.

This does not reproduce with train-2013.03.01. I expect the refactor of user.js/network.js is where the invalidation was dropped.

This, of course, is 5-star and blocks train-2013.03.28.

**Comment by jaredhirsch** [comment]
Hmm. I cannot duplicate _correct_ 123done behavior on dev at all, and I see the `console.log` output that @lloyd added to yahoo.personatest.org/provision when we were debugging the 5-star earlier this week. Kinda suspect that site needs to be redeployed?

@edmoz said he can log in to dev.123done using a yahoo account without problems, so maybe it's just me? Tried clearing cache and clearing yahoo cookies on release channel FF and chrome, still seems busted.

It's close to mfbt at present, will try again in the morning.

**Comment by jaredhirsch** [comment]
I repro'd this and am digging in

**Comment by jaredhirsch** [comment]
Well, I repro'd once. Haven't since. Chrome, FF 19, Safari. No luck.

**Comment by jrgm** [comment]
@6a68 and I worked on this a bit.

To repro in a local dev setup, add these two lines to config/local.json

```
"ephemeral_session_duration_ms": 300000,
"certificate_validity_ms": 300000,
```

which will mean the certificates issued will expire in five minutes (this isn't specific to bigtent or an idp; it shows up for bigtent because it has 5 minute certificates). Then use 127.0.0.1:10001 to signup as a user, wait >5 minutes, and then try to login again. When you fail there is no alert like on 123done.org; On 127.0.0.1 the failure is shown in the page under `logins` section.

Anyways, I found this line, and since we're now using tinyscore instead of underscore, that will throw TypeError
https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/user.js#L67

@6a68 traced through to where this exception is silently eaten. He can add more.

The odd thing is that on Chrome and Safari (maybe) there are mixed results about how reproducible this is. It seems that something/somewhere else can cause this certificate (or client-session) to be cleared and you are re-prompted to authenticate with a password.

So the _(emails.each(...) part is an easy fix, but would like to understand why this is not always reproducible.

**Comment by jaredhirsch** [comment]
@jrgm: I hacked `_(emails).each(...` into `_.each(emails, ...` and pushed to [a branch](https://github.com/6a68/browserid/commits/issue_3153_invalid_certs) and pushed that branch to tinyscorefun.personatest.org. I also pushed your pubkey to that box.

I can't get the bigtent flow to work--I probably configured config.json incorrectly?

At any rate, I couldn't reliably repro yesterday, and can't reliably repro on that box today. Maybe have at it and see if you can repro at all?

**Comment by jaredhirsch** [comment]
@shane-tomlinson So about the apparent bug:
- User.cleanupIdentities [calls the now-undefined method `_(emails)`](https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/user.js#L67)
- `_(emails)` throws a TypeError.
- looks like the TypeError is caught by the inner function, [Network.domainKeyCreationTime](https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/network.js#L556) and converted to an errback
- Unfortunately, the outer function call [drops the errback](https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/user.js#L99) and continues chugging along.

**Comment by jrgm** [comment]
I modified ~app/config.json to add

```
"maximum_event_loop_lag": 1000,
"ephemeral_session_duration_ms": 240000,
"certificate_validity_ms": 240000,
```

which makes this easier to test with just any IDP (I will look at getting this config'd to also do bigtent style which will need to validated, but the issue is independent of bigtent. So uh, not now ;-) ).

Anyways, after forever restartall, I cannot reproduce this in FF (where I could reliably). The other nagging thing in my head is I did also see chrome somehow get out of this trap and I'd like to figure out why (which if it's real is some other issue).
